From b50d247cf28e0862265ae1f5b835769cb133a269 Mon Sep 17 00:00:00 2001
From: Ryan Tomayko <rtomayko@gmail.com>
Date: Sat, 25 Jun 2011 03:02:31 -0700
Subject: [PATCH 1/2] naive hiderefs support for refs/pull/* hierarchy

Don't list refs under the refs/pull/* hierarchy so that
git push --mirror won't try to delete them.

These refs are created on the server when a pull request is created
or modified. A pre-receive hook is configured to reject pushes that
attempt to modify them. People maintain non-github hosted canonical
repos and use git push --mirror to sync with repos on github. The
push --mirror command attempts to delete the refs under refs/pull/*
(the canonical repo does not include them), causing the push to be
rejected.

Omitting these refs from the initial ref advertisement sent back
from receive-pack results in push --mirror not trying to delete
them.
---
 builtin/receive-pack.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/builtin/receive-pack.c b/builtin/receive-pack.c
index e1ba4dc..e7b60cf 100644
--- a/builtin/receive-pack.c
+++ b/builtin/receive-pack.c
@@ -108,6 +108,9 @@ static int receive_pack_config(const char *var, const char *value, void *cb)
 
 static int show_ref(const char *path, const unsigned char *sha1, int flag, void *cb_data)
 {
+	if (strncmp(path, "refs/pull/", strlen("refs/pull/")) == 0)
+		return 0;
+
 	if (sent_capabilities)
 		packet_write(1, "%s %s\n", sha1_to_hex(sha1), path);
 	else
-- 
1.7.5.4


From 4316db861bece57cd2e1d144662127208d773f46 Mon Sep 17 00:00:00 2001
From: Ryan Tomayko <rtomayko@gmail.com>
Date: Sat, 25 Jun 2011 03:20:58 -0700
Subject: [PATCH 2/2] receive.hiderefs config option - takes single prefix

Makes the hidden ref hierarchy path configurable. This is limited to
a single ref prefix. It should probably be broadened to allow a list
of ref prefixes or glob patterns.
---
 builtin/receive-pack.c |    8 +++++++-
 1 files changed, 7 insertions(+), 1 deletions(-)

diff --git a/builtin/receive-pack.c b/builtin/receive-pack.c
index e7b60cf..4bc7814 100644
--- a/builtin/receive-pack.c
+++ b/builtin/receive-pack.c
@@ -34,6 +34,7 @@ static int prefer_ofs_delta = 1;
 static int auto_update_server_info;
 static int auto_gc = 1;
 static const char *head_name;
+static const char *hide_refs;
 static int sent_capabilities;
 
 static enum deny_action parse_deny_action(const char *var, const char *value)
@@ -103,12 +104,17 @@ static int receive_pack_config(const char *var, const char *value, void *cb)
 		return 0;
 	}
 
+	if (strcmp(var, "receive.hiderefs") == 0) {
+		git_config_string(&hide_refs, var, value);
+		return 0;
+	}
+
 	return git_default_config(var, value, cb);
 }
 
 static int show_ref(const char *path, const unsigned char *sha1, int flag, void *cb_data)
 {
-	if (strncmp(path, "refs/pull/", strlen("refs/pull/")) == 0)
+	if (hide_refs && strncmp(path, hide_refs, strlen(hide_refs)) == 0)
 		return 0;
 
 	if (sent_capabilities)
-- 
1.7.5.4

